### PR TITLE
test: add claim submission functional test

### PR DESCRIPTION
**Pull request by warrenv**
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

References https://github.com/poetapp/random/issues/137

**NOTE**: This test will be expanded in the future to use two poet nodes and perform additional
verifications on the claim. This is a first pass to get some functional testing in place since
there currently isn't any. There are also some problems getting integration tests to pass
when specifying a mongodb url via an env var that needs further investigation.

**TO TEST**:
```bash
  $ cp docker-compose.override.yml.example docker-compose.override.yml
  $ docker-compose build
  $ docker-compose up
  $ docker-compose exec poet-node npm run test:functional
```

- Adds an example docker-compose.yml
- Adds a functional test that submits a claim and verifies that certain properties are present on the claim after it gets written to IPFS and batched to the blockchain.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -76,6 +77,7 @@
     "babel-cli": "6.26.0",
     "babel-plugin-module-resolver": "3.1.1",
     "concurrently": "4.0.1",
+    "ipfs-api": "^24.0.1",
```
Comment (geoffturk) [inline package.json]: Pin me!

**Comment by lautarodragan** [comment]
awesome work @warrenv !

I'm trying to test this but some tests are failing:

```
$ dc exec poet-node npm run test:functional

> @po.et/node@1.0.3 test:functional /usr/src/app
> ts-node -r tsconfig-paths/register -r reflect-metadata --files ./tests/functional/claim_with_data.ts

TAP version 13
# submit a claim to a poet node and observe results
ok 1 Given a newly created claim: should be a valid claim
ok 2 Given a claim submission to a poet node: should be accepted
ok 3 Given a claim retrieved by id from poet node A: should have an ipfsFileHash property as a string
ok 4 Given a claim retrieved by id from poet node A: should have an ipfsDirectoryHash property as a string
not ok 5 Given a claim retrieved by id from poet node A: should have a blockHeight property as a number
  ---
    operator: deepEqual
    expected: 'number'
    actual:   'undefined'
    at: assert (/usr/src/app/node_modules/riteway/source/riteway.js:9:14)
    stack: |-
      Error: Given a claim retrieved by id from poet node A: should have a blockHeight property as a number
          at Test.assert [as _assert] (/usr/src/app/node_modules/tape/lib/test.js:224:54)
          at Test.bound [as _assert] (/usr/src/app/node_modules/tape/lib/test.js:76:32)
          at Test.tapeDeepEqual (/usr/src/app/node_modules/tape/lib/test.js:421:10)
          at Test.bound [as same] (/usr/src/app/node_modules/tape/lib/test.js:76:32)
          at assert (/usr/src/app/node_modules/riteway/source/riteway.js:9:14)
          at riteway_1.describe (/usr/src/app/tests/functional/claim_with_data.ts:118:3)
          at process._tickCallback (internal/process/next_tick.js:68:7)
  ...
not ok 6 Given a claim retrieved by id from poet node A: should have a blockHash property as a string
  ---
    operator: deepEqual
    expected: 'string'
    actual:   'undefined'
    at: assert (/usr/src/app/node_modules/riteway/source/riteway.js:9:14)
    stack: |-
      Error: Given a claim retrieved by id from poet node A: should have a blockHash property as a string
          at Test.assert [as _assert] (/usr/src/app/node_modules/tape/lib/test.js:224:54)
          at Test.bound [as _assert] (/usr/src/app/node_modules/tape/lib/test.js:76:32)
          at Test.tapeDeepEqual (/usr/src/app/node_modules/tape/lib/test.js:421:10)
          at Test.bound [as same] (/usr/src/app/node_modules/tape/lib/test.js:76:32)
          at assert (/usr/src/app/node_modules/riteway/source/riteway.js:9:14)
          at riteway_1.describe (/usr/src/app/tests/functional/claim_with_data.ts:125:3)
          at process._tickCallback (internal/process/next_tick.js:68:7)
  ...
ok 7 Given a claim retrieved by id from poet node A: should have a transactionId property as a string

1..7
# tests 7
# pass  5
# fail  2
```

`bitcoin-cli getblockchaininfo | jq ".blocks"` does show an increased number after running the tests, if that helps.

**Comment by warrenv** [comment]
@lautarodragan The environments may be too short for your local environment. Try bumping them up in the `docker-compose.override.yml`.

File: tests/functional/claim_with_data.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  })
+
+  // Wait for a claim batch to be submitted to the blockchain.
+  await pause(parseInt(process.env.TIMESTAMP_INTERVAL_IN_SECONDS, 10))
+
+  // mine N confirmation blocks.
+  await bitcoindClientA.generate(1)
+
+  // Wait for claim batches to be read from the blockchain.
+  await pause(parseInt(process.env.READ_DIRECTORY_INTERVAL_IN_SECONDS, 10))
```
Comment (lautarodragan) [inline tests/functional/claim_with_data.ts]: What do you think of multiplying this value or increasing it, to be on the safer side?

Comment (lautarodragan) [reply tests/functional/claim_with_data.ts]: Nevermind, I just noticed you were doing that already in `pause`.

**Comment by lautarodragan** [comment]
Increasing `TIMESTAMP_INTERVAL_IN_SECONDS` to `10` worked!

File: tests/functional/claim_with_data.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  const claim = await createClaim(privateKey, ClaimType.Work, {
+    name: 'Author Name',
+    text,
+  })
+
+  {
+    const actual = await isValidClaim(claim)
+
+    assert({
```
Comment (kennylavender) [inline tests/functional/claim_with_data.ts]: I don't think this assertion should be part of the test, its helpful for debugging if something does go wrong but I don't see it as part of what this test is responsible for.

Comment (warrenv) [reply tests/functional/claim_with_data.ts]: ok, we'll hash it out in the next round.

**Pull request merged**